blog: Monitoring kubernetes network traffic by using eBPF (#693)
apache/skywalking-website: content/blog/2023-03-18-monitor-kubernetes-network-by-ebpf/kubernetes-service-list.md

In previous articles, we've discussed how to monitor network traffic from programs written in various languages.
This technique remains essential for network traffic monitoring, allowing for the collection of traffic data without language limitations.
However, due to the unique aspects of our monitoring trigger mechanism and the specific features of kernel monitoring, these two areas warrant separate explanations.

### Kernel Monitoring

Kernel monitoring allows users to gain insights into network traffic performance based on the execution at the kernel level,
specifically from Layer 2 (Data Link) to Layer 4 (Transport) of the OSI model.

Network monitoring at the kernel layer is deference from the syscall (user-space) layer in terms of the metrics and identifiers used.
While the syscalls layer can utilize file descriptors to correlate various operations, kernel layer network operations primarily use packets as unique identifiers.
This discrepancy necessitates a mapping relationship that SkyWalking Rover can use to bind these two layers together for comprehensive monitoring.

Let's dive into the details of how data is monitored in both sending and receiving modes.

> #### Sending traffic

Reviewer: ```suggestion
#### Sending Traffic
```

> #### Sending traffic

Reviewer: ```suggestion
#### Observe Sending
```

   This monitoring point is important for assessing the impact of firewall rules and other filtering mechanisms on packet flow.
7. **kprobe/neigh_resolve_output**: If resolving an unknown MAC address is necessary before sending a network request,
   this function records the occurrences and total time spent on this resolution. MAC address resolution times can affect the initial packet transmission delay.
8. **kprobe/__dev_queue_xmit**: Records the start and end times of entering the L2 protocol stack, providing insights into the data link layer's processing times.
9. **tracepoint/net/net_dev_start_xmit** and **tracepoint/net/net_dev_xmit**: Records the actual time taken to transmit each packet at the network interface card (NIC).
   These functions are crucial for understanding the hardware-level performance and potential bottlenecks at the point of sending data to the physical network.

According to the interception of the above method, Apache SkyWalking Rover can provide key execution time and metrics for each level when sending network data,
from the application layer (Layer 7) to the transport layer (Layer 4), and finally to the data link layer (Layer 2).

> #### Receiving traffic

Reviewer: ```suggestion
#### Observe Receiving
```

#### Limitations

The monitoring of stateful protocols like HTTP/2.x currently faces certain limitations:

1. **Inability to Observe Pre-existing Connections**: Monitoring the complete request and response cycle requires that monitoring be initiated before any connections are established.
   This requirement means that connections set up before the start of monitoring cannot be observed.
2. **Challenges with TLS Requests**: Observing TLS encrypted traffic is complex because it relies on **asynchronously attaching uprobes** (user-space attaching) for observation.
   If new requests are made before these uprobes are successfully attached, it becomes impossible to access the data before encryption or after decryption.

## Demo

Next, let’s quickly demonstrate the Kubernetes monitoring feature, so you can understand more specifically what it accomplishes.

### Deploy SkyWalking Showcase

SkyWalking Showcase contains a complete set of example services and can be monitored using SkyWalking. For more information, please check the [official documentation](https://skywalking.apache.org/docs/skywalking-showcase/next/readme/).

> In this demo, we only deploy service, the latest released SkyWalking OAP, and UI.

Reviewer: ```suggestion
In this demo, we only deploy service, the latest(10.0-dev) SkyWalking OAP, and UI.
```

It is not released yet, right? GHCR should be dev images only.

Author: Should we wait for the rover and OAP release steps, and then publish this blog?

Author: Now, I have deleted the image version declaration, please check.

```shell
export SW_OAP_IMAGE=ghcr.io/apache/skywalking/oap
export SW_OAP_IMAGE_TAG=f4836327a65efe76cbbc12f013cc9cbc63a82b2c
export SW_UI_IMAGE=ghcr.io/apache/skywalking/ui
export SW_UI_IMAGE_TAG=f4836327a65efe76cbbc12f013cc9cbc63a82b2c
export SW_ROVER_IMAGE=ghcr.io/apache/skywalking-rover/skywalking-rover:d3a0f874642cf00a5e252440ee3b76a1cea3aac4

export FEATURE_FLAGS=java-agent-injector,single-node,elasticsearch, rover
make deploy.kubernetes

> ```

Reviewer: > Should we wait for the rover and OAP release steps, and then publish this blog?

I think once you update the showcase to the latest, we don't need to export these, and the current and showcase in the future are both working. 
So, we dont' have to wait, right?

Reviewer: Showcase is using the dev versions, rather than released versions.
